packages/x-charts/src/ResponsiveChartContainer/useChartContainerDimensions.ts

```diff
@@ -25,13 +25,32 @@ export const useChartContainerDimensions = (inWidth?: number, inHeight?: number)
 
     setWidth(newWidth);
     setHeight(newHeight);
+
+    return { width: newWidth, height: newHeight };
   }, []);
 
   React.useEffect(() => {
     // Ensure the error detection occurs after the first rendering.
-    displayError.current = true;
+    stateRef.current.displayError = true;
   }, []);
 
+  // This effect is used to compute the size of the container on the initial render.
+  // It is not bound to the raf loop to avoid an unwanted "resize" event.
+  // https://github.com/mui/mui-x/issues/13477#issuecomment-2336634785
+  useEnhancedEffect(() => {
+    // computeRun is used to avoid infinite loops.
+    if (!stateRef.current.initialCompute || stateRef.current.computeRun > 20) {
+      return;
+    }
+
+    const computedSize = computeSize();
+    if (computedSize.width !== width || computedSize.height !== height) {
+      stateRef.current.computeRun += 1;
+    } else if (stateRef.current.initialCompute) {
+      stateRef.current.initialCompute = false;
+    }
+  }, [width, height, computeSize]);
```

Comment:
Could you describe the behavior with for example what is currently happening step by step, and how this PR modify it?

I don't huderstand how adding a `useEnhancedEffect` could help to prevent inifint loop.


If you want a nice reproduction of an infinit loop there is one here:
https://github.com/mui/mui-x/issues/12263#issuecomment-1973343094

Reply:
This useEffect doesn't prevent an infinite loop. Only `computeRun > 20` does. It serves as a circuit breaker in case the layout never settles into a defined size, which would mean this useEffect would run indefinitely. 

This useEffect only happens `before/during` the first render, when the page layout is figuring out what size the components need. Hence it is different from the issues in your example, where the loop in on the render side. 

To fix the loop in your example, we can try to handle it on the `useEnhancedEffect` that has the `ResizeObserver` inside. Though I'm not sure how useful it would be, since my current fix is only for layouts that will definitely settle into a defined width/height, but your example will never settle 😓 

### Explanation of what it does

The main idea of this `useEnhancedEffect` is that everything in it will happen before we see anything, since it is running on the layout effect and committing changes to the state, the state will update synchronously.

The state updates until it settles on a size (new size is the same as last one) or it runs more than 20 times (usually it only requires 5-6 runs from what I tested). 

Once one of these two conditions are met, the hook stops looping (width/height stop changing), and the first actual react render happens.

Reply:
I'm ok with the solution. But this will silently delay the painting without warning the developers about the improvement they could bring. We should add at least a warning if `computeRun > 2`

We can not do `computeRun > 0` because the react strict mode trigger twice the useLayoutEffect. Si the `ref` is updated twice before the state get updated.

Reply:
I had added a warning at first but thought it could be unnecessary and removed it 😅 

Will add it back

Reply:
Hum, I actually removed it because it is an actual valid behaviour that wouldn't require a warning. 🤔 

Like, if the user uses this inside a grid, this will always warn, even though a grid's width would eventually settle in the screen's width.

Reply:
> it is an actual valid behaviour

I disagree on this point. For me the valid behavior is "The container has a defined size". This PR is adding a nice fallback

I tried with Recharts and they have the same behavior as us currently.

https://github.com/user-attachments/assets/f5f67482-1406-4c9a-b863-89075aa62230

From my point of view it would be bad to have such loop without informing user since it silently fix something the developers could potentially do better.

In addition to a warning, it could be a warning plus a prop to silence the warning. Or no warning and a prop to activate the feature

Reply:
As "valid behaviour" I mean it is something the browser does on its own.

I would agree with a prop for this though.

Something like:

```tsx
 /**
  * The chart will try to wait for the parent container to resolve its size
  * before it renders for the first time.
  * 
  * This can be useful in some scenarios where the chart appear to grow after
  * the first render, like when used inside a grid.
  */
  resolveSizeBeforeRender
```